From: Felix Fietkau <nbd@nbd.name>
Date: Fri, 28 Oct 2022 12:26:34 +0200
Subject: [PATCH] flow_dissector: detect DSA using skb->protocol even
 when VLAN tag is present

Fixes flow dissection with ethernet devices that can combine VLAN rx hwaccel
with DSA, since skb->protocol is set to htons(ETH_P_XDSA) by the network stack.

Signed-off-by: Felix Fietkau <nbd@nbd.name>
---

--- a/net/core/flow_dissector.c
+++ b/net/core/flow_dissector.c
@@ -939,7 +939,7 @@ bool __skb_flow_dissect(const struct net
 		hlen = skb_headlen(skb);
 #if IS_ENABLED(CONFIG_NET_DSA)
 		if (unlikely(skb->dev && netdev_uses_dsa(skb->dev) &&
-			     proto == htons(ETH_P_XDSA))) {
+			     skb->protocol == htons(ETH_P_XDSA))) {
 			const struct dsa_device_ops *ops;
 			int offset = 0;
 
